### ISSUE TITLE
demo-transformer runs out of memory

### ISSUE THREAD

**Issue by steelbytes** [androidx/media#758]
`lowmemorykiller: Kill 'androidx.media3.demo.transformer' (2604), uid 10328, oom_score_adj 0 to free 7118060kB rss, 525636kB swap; reason: low watermark is breached and thrashing (210%)`

seems ok for short videos.  my test video is 12GB file from a gopro, dies after a few minutes.

yes I could spend some time trying to debug your app/lib but thought I'd mention first

**Comment by steelbytes** [comment]
steps:
choose large file
video effect choose overlay logo and timer
debug preview off
click export
wait a few minutes
crash

**Comment by steelbytes** [comment]
phone: pixel 7/pro with android 14
vidfeo: 4k@50, 12GB, 15mins

**Comment by droid-girl** [comment]
Thank you for reporting. What kind of overlay are you using? Could you provide image format and the size? Does it take up most of the frame?

**Comment by steelbytes** [comment]
I'm using the demo-transformer app unchanged with logo + timer selected.  so it's overlay which is an the app icon and a bit of text

**Comment by andrewlewis** [comment]
One other follow-up question, do you see the issue if use the exact same setup but with (say) a 180 degree rotation effect (configurable in the demo app) but no overlay? I am wondering if this is specific to applying the overlay or not.

**Comment by steelbytes** [comment]
rotating 180 completes the export BUT then it crashes a few seconds into the playback afterwards due to out of memory exception.

**Comment by dway123** [comment]
I recorded a 28min SDR 4k/60fps video (13GB) on the pixel 7 pro, and tried an export using `demo-transformer` with an overlay plus 180ª rotation. It took 50s to process 30s of video (80s for 130s, 220s for 345s, but basically didn't seem to get slower as it kept going). After some point, the transformation did finish.

After finishing, the video started playback automatically (I touched the exoplayer screen to see the playback controls around this time), and a crash occurred. I'll post more information on an internal bug

**Comment by steelbytes** [comment]
any news?

**Comment by dway123** [comment]
It seems the repro case I thought I had didn't actually result in OOM, but instead the device was running out of disk space (not memory), as the file was ~13GB, and I didn't have 13GB of free space for the muxer's output to save the file. Therefore, the repro case I thought I had isn't exactly the same as yours. That said, I did see some potentially unnecessary Bitmap allocations, which we can hopefully avoid at least 100-150 MB of, on 4k videos.

While I'm not sure I'm able to reproduce the exact failure you had, hopefully reducing our allocations should allow us to reduce memory pressure and avoid the OOM you saw

**Comment by dway123** [comment]
Following up, we've now landed fixes to avoid generating textures and bitmaps on every frame, for text/bitmap/drawable overlays. We never were able to reproduce the exact OOM bug you mentioned, but hopefully this reduces memory pressure and might avoid the OOM. Please sync past [this commit](https://github.com/androidx/media/commit/9add30e5824fda220cae6b6a9b8e53344a9d3512) and let us know if this fixes the OOM you observed. Thanks!

**Comment by steelbytes** [comment]
> Following up, we've now landed fixes to avoid generating textures and bitmaps on every frame, for text/bitmap/drawable overlays. We never were able to reproduce the exact OOM bug you mentioned, but hopefully this reduces memory pressure and might avoid the OOM. Please sync past [this commit](https://github.com/androidx/media/commit/9add30e5824fda220cae6b6a9b8e53344a9d3512) and let us know if this fixes the OOM you observed. Thanks!

still getting oom.

you did say hadn't repeated my test as you ran out of storage.  can you free up some room and try and repeat my test?

**Comment by dway123** [comment]
Sure, I can try the test.

If you don't want to post media publicly please email it, or a link to it, to android-media-github@google.com with subject 'Issue #<issuenumber>' after filing this issue, and note that you will do this here. Otherwise, please send it or a link to it in a comment here directly.

Thanks!

**Comment by steelbytes** [comment]
> Sure, I can try the test.
> 
> If you don't want to post media publicly please email it, or a link to it, to [android-media-github@google.com](mailto:android-media-github@google.com) with subject 'Issue #' after filing this issue, and note that you will do this here. Otherwise, please send it or a link to it in a comment here directly.
> 
> Thanks!

Here's a link

https://1drv.ms/v/s!ArLYtCep2VxVhmUxekocFmlKCYV5?e=mvnU75

**Comment by dway123** [comment]
Sorry, I did try downloading the file using MacOS Chrome+Safari and Linux Chrome+Firefox, and got different errors when trying (`failed to fetch` on MacOS, `unexpected content type null, length 0, errorCodeChain null, internalErrorCode null` on Linux). Could you please try sharing a link as a normal onedrive file instead of onedrive photos, in case that helps?

Or host this file using a different hosting service / let me know how I can download this file. Thanks!

**Comment by steelbytes** [comment]
> Sorry, I did try downloading the file using MacOS Chrome+Safari and Linux Chrome+Firefox, and got different errors when trying (`failed to fetch` on MacOS, `unexpected content type null, length 0, errorCodeChain null, internalErrorCode null` on Linux). Could you please try sharing a link as a normal onedrive file instead of onedrive photos, in case that helps?
> 
> Or host this file using a different hosting service / let me know how I can download this file. Thanks!

Damn.  Ok, I've jsut tried sharing it in a folder direct to you instead of via a read only link.

(onedrive is the only service I have with enough quota)

**Comment by dway123** [comment]
I didn't see an email or link of how this was shared to me, so I'm not sure how to access this. Do you know if you can share this to android-media-github@google.com (or this thread) in a way that generates an email or link?

**Comment by steelbytes** [comment]
> android-media-github@google.com

I had shared to the email shown in your github profile.  I've also now shared to andoird-media-guthib

**Comment by dway123** [comment]
Sorry, I still wasn't able to access this file on OneDrive a few days after the last comment, but forgot to comment here since I figured I'd try again soon after.

I tried again now, and it mentioned "This item might have been deleted, expired, or you might not have permission to access it". I'll just assume I can't access it since this seems a bit difficult, and close as can't repro. Sorry about that...
